### PR TITLE
Add image to column kernel

### PR DESCRIPTION
**Pull request by bartekxk**

### CODE REVIEW

File: example/52_image_to_column/image_to_column_f32.cpp [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+        return false;
+    }
+
+    float ave_time = invoker.Run(argument, StreamConfig{nullptr, config.time_kernel});
+
+    std::size_t num_btype = NDoHoWo * CZYX * sizeof(InDataType);
+
+    float gb_per_sec = num_btype / 1.E6 / ave_time;
+
+    std::cout << "Perf: " << ave_time << " ms, " << gb_per_sec << " GB/s" << std::endl;
```
Comment (aosewski) [inline example/52_image_to_column/image_to_column_f32.cpp]: Please remove those unnecessary empty lines.

File: example/52_image_to_column/image_to_column_f32.cpp [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+        std::cerr << "wrong! device_img2col with the specified compilation parameters does "
+                     "not support this img2col problem"
+                  << std::endl;
+
+        return false;
+    }
+
+    float ave_time = invoker.Run(argument, StreamConfig{nullptr, config.time_kernel});
+
+    std::size_t num_btype = NDoHoWo * CZYX * sizeof(InDataType);
```
Comment (aosewski) [inline example/52_image_to_column/image_to_column_f32.cpp]: Shouldn't this be multiplied by 2 here? You read and then write the same amount of data.

File: example/52_image_to_column/image_to_column_f32.cpp [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+    ck::utils::conv::ConvParam conv_params = DefaultConvParams;
+
+    if(!parse_cmd_args(argc, argv, config, conv_params))
+    {
+        return EXIT_FAILURE;
+    }
+
+    if(conv_params.num_dim_spatial_ != NDimSpatial)
+    {
+        std::cerr << "unsupported # of spatials dimensions" << std::endl;
```
Comment (aosewski) [inline example/52_image_to_column/image_to_column_f32.cpp]:
```suggestion
        std::cerr << "unsupported # of spatial dimensions" << std::endl;
```

File: include/ck/tensor_operation/gpu/device/device_image_to_column.hpp [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+#include "ck/tensor_operation/gpu/device/device_base.hpp"
+
+namespace ck {
+namespace tensor_operation {
+namespace device {
+
+// Image to column:
+//   input : input image [N, Di, Hi, Wi, C],
+//   output : output image [N * Do * Ho * Wo, Z *  Y * X * C]
```
Comment (aosewski) [inline include/ck/tensor_operation/gpu/device/device_image_to_column.hpp]: You could make it a doxygen style doc.

File: library/src/tensor_operation_instance/gpu/image_to_column/device_image_to_column_nhwc_1d_instance.cpp [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    std::vector<std::unique_ptr<DeviceImageToColumn<1, GNWC, F32, F32>>>& instances)
+{
+    add_device_operation_instances(instances, device_image_to_column_f32_instances<1, GNWC>{});
+}
+
+void add_device_image_to_column_nhwc_1d_i8_instances(
+    std::vector<std::unique_ptr<DeviceImageToColumn<1, GNWC, int8_t, int8_t>>>& instances)
+{
+    add_device_operation_instances(instances, device_image_to_column_i8_instances<1, GNWC>{});
+}
```
Comment (aosewski) [inline library/src/tensor_operation_instance/gpu/image_to_column/device_image_to_column_nhwc_1d_instance.cpp:34]: Do we really need instances for all those data types? @zjing14

File: library/include/ck/library/reference_tensor_operation/cpu/reference_image_to_column.hpp [hunk 6]
```diff
@@ -0,0 +1,327 @@
...
+
+namespace ck {
+namespace tensor_operation {
+namespace host {
+
+//
+// @brief      Reference implementation for image to column.
+//
+// @paragraph
+//             Tensor descriptor in NCHW dimensional order
```
Comment (aosewski) [inline library/include/ck/library/reference_tensor_operation/cpu/reference_image_to_column.hpp]:
```suggestion
//             Tensor descriptor has NCHW data layout.
```

File: library/include/ck/library/reference_tensor_operation/cpu/reference_image_to_column.hpp [hunk 7]
```diff
@@ -0,0 +1,327 @@
...
+//
+// @tparam     InDataType               Input tensor data type.
+// @tparam     OutDataType              Output tensor data type.
+// @tparam     NDimSpatial  Number of spatial dimensions.
+//
+// input descriptor in [N, C, Di, Hi, Wi] order
+// output descriptor in [N * Do * Ho * Wo, C * Z * Y * X] order
+// phyiscal layout is [N, Di, Hi, Wi, C]
```
Comment (aosewski) [inline library/include/ck/library/reference_tensor_operation/cpu/reference_image_to_column.hpp]:
```suggestion
// memory layout is [N, Di, Hi, Wi, C]
```

File: library/include/ck/library/reference_tensor_operation/cpu/reference_image_to_column.hpp [hunk 8]
```diff
@@ -0,0 +1,327 @@
...
+        {
+            return Run(*dynamic_cast<const Argument*>(p_arg));
+        }
+    };
+
+    static constexpr bool IsValidCompilationParameter()
+    {
+        using namespace tensor_layout::convolution;
+
+        if(!(std::is_same_v<InputLayout, GNWC> || std::is_same_v<InputLayout, GNHWC> ||
```
Comment (aosewski) [inline library/include/ck/library/reference_tensor_operation/cpu/reference_image_to_column.hpp]: `if constexpr`

File: library/include/ck/library/reference_tensor_operation/cpu/reference_image_to_column.hpp [hunk 9]
```diff
@@ -0,0 +1,327 @@
...
+             std::is_same_v<InputLayout, GNDHWC>))
+        {
+            return false;
+        }
+        if(!(NDimSpatial >= 1 && NDimSpatial <= 3))
```
Comment (aosewski) [inline library/include/ck/library/reference_tensor_operation/cpu/reference_image_to_column.hpp]: `if constexpr`

File: include/ck/tensor_operation/gpu/device/impl/device_image_to_column_impl.hpp [hunk 10]
```diff
@@ -0,0 +1,405 @@
...
+
+#pragma once
+
+#include "ck/tensor_description/tensor_descriptor.hpp"
+#include "ck/tensor_description/tensor_descriptor_helper.hpp"
+#include "ck/tensor_operation/gpu/device/device_image_to_column.hpp"
+#include "ck/tensor_operation/gpu/grid/gridwise_image_to_column.hpp"
+#include "ck/host_utility/device_prop.hpp"
+#include "ck/host_utility/kernel_launch.hpp"
+#include "ck/host_utility/stream_utility.hpp"
```
Comment (aosewski) [inline include/ck/tensor_operation/gpu/device/impl/device_image_to_column_impl.hpp]: Not used.

File: include/ck/tensor_operation/gpu/device/impl/device_image_to_column_impl.hpp [hunk 11]
```diff
@@ -0,0 +1,405 @@
...
+        float Run(const Argument& arg, const StreamConfig& stream_config = StreamConfig{})
+        {
+            if(stream_config.log_level_ > 0)
+            {
+                arg.Print();
+            }
+
+            const auto block_2_tile_map =
+                BlockToCTileMap_M00_N0_M01Adapt<MPerBlock, KPerBlock, OutputGridDesc>(
+                    arg.out_grid_desc_m_k_, I1 /*M01*/);
```
Comment (aosewski) [inline include/ck/tensor_operation/gpu/device/impl/device_image_to_column_impl.hpp]: Why you choose `M01` equal to `1` ? If I remember correctly by default it is `8`. Recently there was a PR from PoYen Chen, where he benchmarked perf wrt this parameter value. You can take a look.

File: include/ck/tensor_operation/gpu/device/impl/device_image_to_column_impl.hpp [hunk 12]
```diff
@@ -0,0 +1,405 @@
...
+        if(!(NDimSpatial >= 1 && NDimSpatial <= 3))
+        {
+            return false;
+        }
+
+        const auto x_pad_left  = arg.input_left_pads_[NDimSpatial - I1];
+        const auto x_pad_right = arg.input_right_pads_[NDimSpatial - I1];
+        const auto dilation_x  = arg.conv_filter_dilations_[NDimSpatial - I1];
+        const auto stride_x    = arg.conv_filter_strides_[NDimSpatial - I1];
+        bool is_c_packed       = arg.input_g_n_c_wis_strides_[NDimSpatial + I2] == arg.C_;
```
Comment (aosewski) [inline include/ck/tensor_operation/gpu/device/impl/device_image_to_column_impl.hpp]: If the input strides are indeed in `g, n, c, wis.. ` layout then  `[NDimSpatial + I2]` index may access C. But if NDimSpatial is > 0 then you access spatial dims.

File: include/ck/tensor_operation/gpu/grid/gridwise_image_to_column.hpp [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+#pragma once
+
+#include "ck/utility/common_header.hpp"
+#include "ck/tensor_description/multi_index_transform_helper.hpp"
+#include "ck/tensor_description/tensor_descriptor.hpp"
+#include "ck/tensor_description/tensor_descriptor_helper.hpp"
+#include "ck/tensor_operation/gpu/grid/block_to_ctile_map.hpp"
+#include "ck/tensor_operation/gpu/grid/gridwise_gemm_pipeline_selector.hpp"
+#include "ck/tensor_operation/gpu/block/blockwise_gemm_xdlops.hpp"
+#include "ck/tensor_operation/gpu/block/thread_group_tensor_slice_transfer_v4r1.hpp"
```
Comment (aosewski) [inline include/ck/tensor_operation/gpu/grid/gridwise_image_to_column.hpp]: Unused.

File: include/ck/tensor_operation/gpu/grid/gridwise_image_to_column.hpp [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+
+        copy_global_to_global.Run(
+            tie(in_grid_desc), tie(in_global_buf), tie(out_grid_desc), tie(out_global_buf));
+    }
+
+    __host__ static constexpr bool CheckValidity(const InputGridDesc& in_grid_desc,
+                                                 const OutputGridDesc& out_grid_desc)
+    {
+        if(in_grid_desc.GetLength(I0) % MPerBlock != 0 ||
+           in_grid_desc.GetLength(I1) % KPerBlock != 0)
```
Comment (aosewski) [inline include/ck/tensor_operation/gpu/grid/gridwise_image_to_column.hpp:89]: Shouldn't here be a logical conjunction instead of an alternative ?

**Review by aosewski — COMMENTED** [review]
I'll finish tomorrow profiler and tests review.

File: profiler/include/profiler/profile_image_to_column_impl.hpp [hunk 15]
```diff
@@ -0,0 +1,213 @@
...
+#include <limits>
+
+#include "ck/ck.hpp"
+#include "ck/tensor_operation/gpu/device/tensor_layout.hpp"
+#include "ck/tensor_operation/gpu/device/device_image_to_column.hpp"
+
+#include "ck/tensor_operation/gpu/device/impl/device_image_to_column_impl.hpp"
+
+#include "ck/library/tensor_operation_instance/gpu/image_to_column.hpp"
+
```
Comment (aosewski) [inline profiler/include/profiler/profile_image_to_column_impl.hpp]: Why those empty lines?

File: profiler/include/profiler/profile_image_to_column_impl.hpp [hunk 16]
```diff
@@ -0,0 +1,213 @@
...
+    // get device op instances
+    const auto op_ptrs = ck::tensor_operation::device::instance::DeviceOperationInstanceFactory<
+        DeviceOp>::GetInstances();
+
+    std::cout << "found " << op_ptrs.size() << " instances" << std::endl;
+
+    std::string best_op_name;
+    float best_avg_time = std::numeric_limits<float>::max();
+    ;
```
Comment (aosewski) [inline profiler/include/profiler/profile_image_to_column_impl.hpp]: Leftover

File: profiler/include/profiler/profile_image_to_column_impl.hpp [hunk 17]
```diff
@@ -0,0 +1,213 @@
...
+            auto invoker_ptr = op_ptr->MakeInvokerPointer();
+
+            float avg_time =
+                invoker_ptr->Run(argument_ptr.get(), StreamConfig{nullptr, time_kernel});
+
+            std::size_t num_btype =
+                conv_param.GetInputByte<InputDataType>() + NDoHoWo * CZYX * sizeof(OutputDataType);
+
+            float gb_per_sec = num_btype / 1.E6 / avg_time;
+
```
Comment (aosewski) [inline profiler/include/profiler/profile_image_to_column_impl.hpp]: Could you please remove all those empty lines?

File: profiler/include/profiler/profile_image_to_column_impl.hpp [hunk 18]
```diff
@@ -0,0 +1,213 @@
...
+            std::string op_name = op_ptr->GetTypeString();
+
+            auto invoker_ptr = op_ptr->MakeInvokerPointer();
+
+            float avg_time =
+                invoker_ptr->Run(argument_ptr.get(), StreamConfig{nullptr, time_kernel});
+
+            std::size_t num_btype =
+                conv_param.GetInputByte<InputDataType>() + NDoHoWo * CZYX * sizeof(OutputDataType);
```
Comment (aosewski) [inline profiler/include/profiler/profile_image_to_column_impl.hpp]: Isn't it incorrect? You read from input multiple times the same elements. You read `NDoHoWo * CZYX` number of input elements.

File: profiler/include/profiler/profile_image_to_column_impl.hpp [hunk 19]
```diff
@@ -0,0 +1,213 @@
...
+    // get device op instances
+    const auto op_ptrs = ck::tensor_operation::device::instance::DeviceOperationInstanceFactory<
+        DeviceOp>::GetInstances();
+
+    std::cout << "found " << op_ptrs.size() << " instances" << std::endl;
+
+    std::string best_op_name;
+    float best_avg_time = std::numeric_limits<float>::max();
+    ;
+    float best_tflops     = 0;
```
Comment (aosewski) [inline profiler/include/profiler/profile_image_to_column_impl.hpp]: Unused ?

File: profiler/include/profiler/profile_image_to_column_impl.hpp [hunk 20]
```diff
@@ -0,0 +1,213 @@
...
+        }
+        else
+        {
+            std::cout << op_ptr->GetTypeString() << " does not support this problem" << std::endl;
+        }
+    }
+
+    std::cout << "Best configuration parameters:"
+              << "\nname: " << best_op_name << "\navg_time: " << best_avg_time
+              << "\ntflops: " << best_tflops << "\nGB/s: " << best_gb_per_sec << std::endl;
```
Comment (aosewski) [inline profiler/include/profiler/profile_image_to_column_impl.hpp]: You do not calculate tflops, so there's no point in showing this information.

File: profiler/include/profiler/profile_image_to_column_impl.hpp [hunk 21]
```diff
@@ -0,0 +1,213 @@
...
+
+    return is_supporting_instance && pass;
```
Comment (aosewski) [inline profiler/include/profiler/profile_image_to_column_impl.hpp:196]: :+1:

**Review by aosewski — COMMENTED** [review]
In general: great work!

File: example/52_image_to_column/image_to_column_f32.cpp (hunk 1 above)
Comment (bartekxk) [reply example/52_image_to_column/image_to_column_f32.cpp]: Done.

File: example/52_image_to_column/image_to_column_f32.cpp (hunk 2 above)
Comment (bartekxk) [reply example/52_image_to_column/image_to_column_f32.cpp]: Done.  You are righht.

File: example/52_image_to_column/image_to_column_f32.cpp (hunk 3 above)
Comment (bartekxk) [reply example/52_image_to_column/image_to_column_f32.cpp]: Done.

File: include/ck/tensor_operation/gpu/device/device_image_to_column.hpp (hunk 4 above)
Comment (bartekxk) [reply include/ck/tensor_operation/gpu/device/device_image_to_column.hpp]: Done.

File: library/include/ck/library/reference_tensor_operation/cpu/reference_image_to_column.hpp (hunk 6 above)
Comment (bartekxk) [reply library/include/ck/library/reference_tensor_operation/cpu/reference_image_to_column.hpp]: Done.

File: library/include/ck/library/reference_tensor_operation/cpu/reference_image_to_column.hpp (hunk 7 above)
Comment (bartekxk) [reply library/include/ck/library/reference_tensor_operation/cpu/reference_image_to_column.hpp]: Done.

File: library/include/ck/library/reference_tensor_operation/cpu/reference_image_to_column.hpp (hunk 8 above)
Comment (bartekxk) [reply library/include/ck/library/reference_tensor_operation/cpu/reference_image_to_column.hpp]: Done.

File: library/include/ck/library/reference_tensor_operation/cpu/reference_image_to_column.hpp (hunk 9 above)
Comment (bartekxk) [reply library/include/ck/library/reference_tensor_operation/cpu/reference_image_to_column.hpp]: Done.

File: include/ck/tensor_operation/gpu/device/impl/device_image_to_column_impl.hpp (hunk 12 above)
Comment (bartekxk) [reply include/ck/tensor_operation/gpu/device/impl/device_image_to_column_impl.hpp]: Done. Checked also for w_packed.

File: include/ck/tensor_operation/gpu/grid/gridwise_image_to_column.hpp (hunk 13 above)
Comment (bartekxk) [reply include/ck/tensor_operation/gpu/grid/gridwise_image_to_column.hpp]: Removed.

File: profiler/include/profiler/profile_image_to_column_impl.hpp (hunk 19 above)
Comment (bartekxk) [reply profiler/include/profiler/profile_image_to_column_impl.hpp]: Done.

File: profiler/include/profiler/profile_image_to_column_impl.hpp (hunk 20 above)
Comment (bartekxk) [reply profiler/include/profiler/profile_image_to_column_impl.hpp]: Removed.

File: profiler/include/profiler/profile_image_to_column_impl.hpp (hunk 18 above)
Comment (bartekxk) [reply profiler/include/profiler/profile_image_to_column_impl.hpp]: Fixed.

File: profiler/include/profiler/profile_image_to_column_impl.hpp (hunk 17 above)
Comment (bartekxk) [reply profiler/include/profiler/profile_image_to_column_impl.hpp]: Done.

File: profiler/include/profiler/profile_image_to_column_impl.hpp (hunk 16 above)
Comment (bartekxk) [reply profiler/include/profiler/profile_image_to_column_impl.hpp]: Removed.

File: profiler/include/profiler/profile_image_to_column_impl.hpp (hunk 15 above)
Comment (bartekxk) [reply profiler/include/profiler/profile_image_to_column_impl.hpp]: Removed.

File: include/ck/tensor_operation/gpu/device/impl/device_image_to_column_impl.hpp (hunk 10 above)
Comment (bartekxk) [reply include/ck/tensor_operation/gpu/device/impl/device_image_to_column_impl.hpp]: Removed.

File: library/src/tensor_operation_instance/gpu/image_to_column/device_image_to_column_nhwc_1d_instance.cpp (hunk 5 above)
Comment (bartekxk) [reply library/src/tensor_operation_instance/gpu/image_to_column/device_image_to_column_nhwc_1d_instance.cpp:34]: We can check but compilation of one of the instance file takes only approx. 4% of whole compilation time per thread (we have three files for each ndimspatial) .  I tested it for 32 compilation threads.

File: include/ck/tensor_operation/gpu/device/impl/device_image_to_column_impl.hpp (hunk 11 above)
Comment (bartekxk) [reply include/ck/tensor_operation/gpu/device/impl/device_image_to_column_impl.hpp]: I didnt noticed any perf change but I applied Your suggestion.

**Review by aosewski — APPROVED** [review]

File: library/src/tensor_operation_instance/gpu/image_to_column/device_image_to_column_nhwc_1d_instance.cpp (hunk 5 above)
Comment (zjing14) [reply library/src/tensor_operation_instance/gpu/image_to_column/device_image_to_column_nhwc_1d_instance.cpp:34]: > Do we really need instances for all those data types? @zjing14

Yes.

**Pull request merged**